### ISSUE TITLE
sqs-> lambda -> sqs -> lambda - event source mapping problem

### ISSUE THREAD

**Issue by Rafal-Laskowski** [localstack/localstack#2198]
# Type of request: This is a ...

[X] bug report

# Detailed description
I am a QA in a team and I have an AWS system. I recreated it on localstack for testing purposes.
I need to be able to preserve SQS messages for automated tests and I can't interfere with existing components.

I have an SQS which is connected to Lambda with event-source-mapping. To preserve SQS messages I've created a logic to create proxy-lambda and proxy-sqs and put them in between original SQS and lambda.

My proxy lambda will save the records to DynamoDB but it's not important for this PoC.

The flow looks like this:

Initial SQS -> My Proxy Lambda -> My Proxy SQS -> Final Lambda

My Proxy Lambda saved the messages in DynamoDB and passes them to My Proxy SQS which, I, as far as I know, should invoke Final Lambda with a message because they are connected with event-source-mapping

## Expected behaviour
My proxy-sqs should invoke final lambda since they are connected with event-source-mapping

## Actual behaviour
proxy-sqs gets the message but doesn't invoke lambda connected with event-source-mapping

# Steps to reproduce

Build a system with the following:
Initial SQS (from the original system)
Second SQS
Proxy Lambda
Final Lambda (from the original system)
Connect Initial SQS with Proxy Lambda with event-source-mapping
Connect Second SQS to Final Lambda with event-source-mapping
Send SQS message to Initial SQS
Final Lambda should receive the message but it's in the queue of Second SQS

## Command used to start LocalStack
Start localstack with `LAMBDA_EXECUTOR=local` and `SERVICES=dynamodb,sns,sqs,s3,lambda,cloudwatch`

docker-compose.yml:
```
version: '2.1'

services:
  localstack:
    image: localstack/localstack:0.10.7
    ports:
      - "4567-4597:4567-4597"
      - "${PORT_WEB_UI-18080}:${PORT_WEB_UI-18080}"
    environment:
      - SERVICES=${SERVICES-dynamodb,sns,sqs,s3,lambda,cloudwatch}
      - DEBUG=${DEBUG- }
      - DATA_DIR=${DATA_DIR- }
      - PORT_WEB_UI=${PORT_WEB_UI- }
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-local}
      - KINESIS_ERROR_PROBABILITY=${KINESIS_ERROR_PROBABILITY- }
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "${TMPDIR:-/tmp/local-stack}:/tmp/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
```

### Step 1 - Create initial sqs

```
aws --endpoint-url=http://localhost:4576 sqs create-queue --queue-name initial_sqs
```

### Step 2 - Create second sqs
```
aws --endpoint-url=http://localhost:4576 sqs create-queue --queue-name second_sqs
```

### Step 3 - Create lambda
AWS CLI command:
```
aws --endpoint-url=http://localhost:4574 lambda create-function --function-name proxy-lambda --runtime python3.7 --role arn:aws:lambda:us-east-1:000000000000 --handler lambda_function.lambda_handler --zip-file "fileb://lambda.zip" --environment "Variables={TARGET_DATABASE=my_table,TARGET_SQS=second_sqs}" --timeout 60
```
Python code:
```
import logging
import os
from datetime import datetime

import boto3

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)


def get_local_resource(service, port):
    return boto3.resource(service, endpoint_url='http://localhost:{}'.format(port), region_name='us-east-1')


def get_sqs():
    return get_local_resource('sqs', '4576')


def get_dynamodb():
    return get_local_resource('dynamodb', '4569')


def lambda_handler(event, context):
    logger.info("Preserve messages lambda invoke")
    records = event.get('Records')

    if not records:
        return

    send_messages_to_fake_sqs(records)


def send_messages_to_fake_sqs(records):
    sqs_queue_name = os.environ['TARGET_SQS']
    sqs = get_sqs()
    sqs_queue = sqs.get_queue_by_name(QueueName=sqs_queue_name)

    entries = []
    i = 1

    for record in records:
        entries.append({
            'Id': str(i),
            'MessageBody': str(record.get('body'))
        })
        i += 1

    response = sqs_queue.send_messages(Entries=entries)

    failed_response = response.get('Failed')
    if not failed_response:
        logger.info("Forwarded {} messages to sqs: {}".format(len(entries), sqs_queue_name))
        logger.info(response.get('Successful'))
    else:
        logger.error(failed_response)
```

### Step 4 - Create event-source-mapping between `proxy-lambda` and `initial_sqs`
```
aws --endpoint-url=http://localhost:4574 lambda create-event-source-mapping --function-name proxy-lambda --batch-size 5 --event-source-arn arn:aws:sqs:us-east-1:000000000000:initial_sqs
```

### Step 5 - Create final lambda
AWS CLI Command:
```
aws --endpoint-url=http://localhost:4574 lambda create-function --function-name final_lambda --runtime python3.7 --role arn:aws:lambda:us-east-1:000000000000 --handler lambda_function.lambda_handler --zip-file "fileb://lambda.zip" --timeout 60
```

Python Code:
```
import logging

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)


def lambda_handler(event, context):
    logger.info("FINAL LAMBDA")
```

### Step 6 - Create event-source-mapping between `final_lambda` and `second_sqs`
```
aws --endpoint-url=http://localhost:4574 lambda create-event-source-mapping --function-name final_lambda --batch-size 5 --event-source-arn arn:aws:sqs:us-east-1:000000000000:second_sqs
```

### Step 7 - Send message to `initial_sqs`
```
aws --endpoint-url=http://localhost:4576 sqs send-message --queue-url http://localhost:4576/queue/initial_sqs --message-body "Hello world!"
```

### Step 8 - Read logs of `proxy-lambda` to make sure it got the message and forwarded it
```
aws --endpoint-url=http://localhost:4586 logs describe-log-streams --log-group-name /aws/lambda/proxy-lambda
```
```
aws --endpoint-url=http://localhost:4586 logs get-log-events --log-group-name /aws/lambda/proxy-lambda --log-stream-name ???
```

### Step 9 - Check the logs of `final_lambda` and verify that there is no lambda invocation or logs
```
aws --endpoint-url=http://localhost:4586 logs describe-log-streams --log-group-name /aws/lambda/final_lambda
```

### Step 10 - Make sure that the message reached `second_sqs`
```
aws --endpoint-url=http://localhost:4576 sqs receive-message --queue-url http://localhost:4576/queue/second_sqs
```

**Comment by mrosales** [comment]
Its likely that this is caused because the sqs -> lambda trigger is only invoked for the `SendMessage` API call, not `SendMessageBatch`. See: #1364

**Comment by Rafal-Laskowski** [comment]
I think we can close the issue due to merged fix in #1364

**Comment by whummer** [comment]
Great, should be fixed in #2215 . Thanks again!
